### PR TITLE
feat(reputation): implement a dynamic threshold for reppoe

### PR DESCRIPTION
**Pull request by lrubiorod**
Co-authored-by: Tomasz Polaczyk <tmpolaczyk@gmail.com>

### CODE REVIEW

**Comment by tmpolaczyk** [comment]
We can re-use the cache to remove calls to

    rep_eng.trs().get_sum(rep_eng.ars().active_identities())

File: validations/benches/reppoe.rs [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+use witnet_validations::validations;
+// To benchmark the old algorithm, comment out the line that says cfg any:
+#[cfg(any())]
+mod validations {
+    use super::*;
+
+    pub fn calculate_reppoe_threshold(
+        rep_eng: &ReputationEngine,
+        pkh: &PublicKeyHash,
+        _dr_pkh: &PublicKeyHash,
```
Comment (lrubiorod) [inline validations/benches/reppoe.rs]: ??

File: validations/benches/reppoe.rs [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        let pkh = pkh_i(i as u32);
+        rep_eng.ars_mut().push_activity(vec![pkh]);
+        rep_eng
+            .trs_mut()
+            .gain(Alpha(10), vec![(pkh, Reputation(rep))])
+            .unwrap();
+    }
+    // Initialize cache
+    rep_eng.total_active_reputation();
+    //validations::calculate_reppoe_threshold(&rep_eng, &my_pkh, &dr_pkh, num_witnesses);
```
Comment (lrubiorod) [inline validations/benches/reppoe.rs]: ??

File: validations/benches/reppoe.rs (hunk 1 above)
Comment (tmpolaczyk) [reply validations/benches/reppoe.rs]: :man_shrugging:

File: data_structures/src/chain.rs [hunk 3]
```diff
@@ -36,6 +36,7 @@ use crate::{
     vrf::{BlockEligibilityClaim, DataRequestEligibilityClaim},
 };
 use bech32::{FromBase32, ToBase32};
+use std::cell::RefCell;
```
Comment (tmpolaczyk) [inline data_structures/src/chain.rs]: fmt

File: data_structures/src/chain.rs [hunk 4]
```diff
@@ -1854,8 +1858,154 @@ impl ReputationEngine {
...
+        if !self.valid {
+            let (total_active_rep, sorted_active_rep) = gen();
+            self.initialize(total_active_rep, sorted_active_rep);
+        }
+
+        self.total_active_rep
+    }
+}
+
+/// Recursive function for the threshold_factor function
```
Comment (tmpolaczyk) [inline data_structures/src/chain.rs]: Not recursive anymore?

**Pull request merged**